Feature/search multiple tags

mosegontar opened the pull request:
This PR implements the "Search multiple tags, exclusion in tag search" feature, referenced in #186.

It adds support for finding bookmarks that match 

1) ANY tag in a set of tags: `buku --stag tag1, tag2, ... tagn`
2) ALL tags in a set of tags: `buku --stag tag1 + tag2 + tag3 ... + tagn`

It also adds support for excluding bookmarks that match certain tags:

`buku --stag tag2, tag3 - tag1`

will return bookmarks that match either tag2 or tag3, but will exclude from the results bookmarks that match tag1.

I've added test cases to `tests/test_bukuDb.py` for each of the three scenarios above.

#### Program help

I've updated the following sections of the README: Cmndline options, Operational notes, examples (19-21).

I've updated the man page's program options. I may have done the notation/syntax describing the the options wrong, so please let me know and I'll amend!

Please let me know if anything should change. Thanks!

jarun commented on README.md:
```diff
@@ -283,7 +286,7 @@ PROMPT KEYS:
   - --deep : match **substrings** (`match` matches `rematched`) in URL, title and tags.
   - --sreg : match a regular expression (ignores --deep).
-  - --stag : search bookmarks by a tag, or list all tags alphabetically with usage count (if no arguments).
+  - --stag : search bookmarks by tags, or list all tags alphabetically with usage count (if no arguments). Delimit the list of tags in the query with `,` to search for bookmarks that match ANY of the listed tags. Delimit tags with `+` to search for bookmarks that match ALL of the listed tags. (Note: these cannot be combined). Exclude bookmarks matching certain tags from the results by using `+` f ...
```
> Exclude bookmarks matching certain tags from the results by using `+` followed by the tags.

should be

> Exclude bookmarks matching certain tags from the results by using `-` followed by the tags.

mosegontar replied:
This should be all set.

jarun commented on buku.py:
```diff
@@ -2042,7 +2068,7 @@ def prompt_help(file=sys.stdout):
     d                      match substrings ('pen' matches 'opened')
     r expression           run a regex search
-    t [...]                search bookmarks by a tag or show taglist
+    t [...]                search bookmarks by tags or show tag list
```
`taglist`

mosegontar replied:
All set.

jarun commented:
The core logic looks fine. Also, please update the man page with the option improvements and examples.

jarun commented on buku.py:
```diff
@@ -1079,18 +1079,44 @@ def searchdb(self, keywords, all_keywords=False, deep=False, regex=False):
-        logdbg('query: "%s", args: %s', query, tag)
+        # do not allow combination of search logics
+        if '+' in tags and ',' in tags:
+            print('Cannot use both "+" and "," in same search')
```
Please change to

> logerr("Cannot use both '+' and ',' in same search")

mosegontar replied:
All set.

jarun commented:
I guess we must document `-` should be space separated:

```
~$ ./buku.py -t c program execution,interview-programming
1. :: DeveloperIQ.in - Asia's #1 Software Technology magazine :: [102]
   > http://developeriq.in/articles/2012/jan/06/understanding-the-execution-of-a-c-program/
   # c program execution,interview,programming

2. Linux x86 Program Start Up [103]
   > http://dbp-consulting.com/tutorials/debugging/linuxProgramStartup.html
   # c program execution,interview,programming

buku (? for help)
```

and

```
~$ ./buku.py -t c program execution,interview - programming
// many results having any of `c program execution` or `interview` tags 
```

and

```
~$ ./buku.py -t c program execution+interview - programming
~$
```

but

```
~$ ./buku.py -t c program execution+interview+programming
1. :: DeveloperIQ.in - Asia's #1 Software Technology magazine :: [102]
   > http://developeriq.in/articles/2012/jan/06/understanding-the-execution-of-a-c-program/
   # c program execution,interview,programming

2. Linux x86 Program Start Up [103]
   > http://dbp-consulting.com/tutorials/debugging/linuxProgramStartup.html
   # c program execution,interview,programming

buku (? for help)
```

jarun commented:
You mentioned you have updated the man page but it's missing in the changeset.

mosegontar commented:
Sorry that was confusion on my part with respect to the man page (apologies I'm pretty green)! It's updated now.

mosegontar commented:
Regarding documenting that `-` should be space separated, I've gone ahead and done this in the program options, README, and man page. 

Thanks for raising the concern about space separation of `+`. I'm inclined to enforce space separation of `+` in `search_by_tag` because current users may have tags with `+` in them.

I do like the idea of treating `+` in the same manner as `,` (since their functions are similar in `search_by_tag`), so space separation would not be needed. But given the concern about current users, I think space separation is probably preferable and the least disruptive change.

Let me know what you think. If you agree, I can make updates to `search_by_tag` and document accordingly. Thanks a lot!

jarun commented:
Yes, we should enforce space separation for both + and -. We do that for specific tag append, removal too. Let's keep these uniform and treat DELIM (comma) as a special case.

mosegontar commented:
I've updated `search_by_tag` to enforce space separation for `+`. Also updated README and man page with notes about this constraint. (Though I'm failing the build now, I'm not sure why).

But it did just occur to me that there might be a better/simpler way to handle this. We keep 'search any' the same as we have it now (`buku --stag tag1, tag2`). But to 'search all', we can require that a single `+` option be set immediately after `--stag`. So 'search all' would look like 

`buku --stag + tag1, tag2`

This is along the same lines as how +/- are used with `--tag` to append or remove a tag from the tagset. Another benefit is that I think it's more natural to list out your tags as with commas. What do you think?

jarun commented:
There's a build failure:

```
4.09s$ python3 -m flake8
./tests/test_bukuDb.py:412:21: E128 continuation line under-indented for visual indent
./tests/test_bukuDb.py:417:21: E128 continuation line under-indented for visual indent
The command "python3 -m flake8" exited with 1.
```

Please fix.

jarun commented:
@rachmadaniHaryono this is a sizable PR. Can you please review the changes as well? Thanks!

mosegontar commented:
Ah, fixed under-indent causing the build failure.

jarun commented:
Thank you @mosegontar! I'll wait for @rachmadaniHaryono to complete the review.

In the meantime, if you are interested, please let me know if you would like to join the team as a collaborator.

rachmadaniHaryono commented on tests/test_bukuDb.py:
```diff
@@ -274,6 +274,174 @@ def test_search_by_tag(self):
+            ]
+            self.assertEqual(results, expected)
+
+    # @unittest.skip('skipping')
```
remove this if not really needed

rachmadaniHaryono commented:
```# @unittest.skip('skipping')``` that can be removed, unless it is necessary 

i am not very famliar with db command, and i haven't really checked if the method is correct, so please check the db query. maybe test debug message if and give example for each case

also from single ```tag = delim_wrap(tag.strip(DELIM))```to these lines [1089](https://github.com/jarun/Buku/pull/187/files#diff-3e8df16046c4f0c0ecf1a4ae165fe955R1089) - [1106](https://github.com/jarun/Buku/pull/187/files#diff-3e8df16046c4f0c0ecf1a4ae165fe955R1106) is quite hard to follow.

mosegontar commented:
@jarun oh thanks a lot. yes, I am interested! I'll write to you on gitter.

jarun commented:
Sure! I'll add you.

@rachmadaniHaryono please let us know if we are good to go with the latest changes.

jarun commented:
@mosegontar I've sent you the invite.

mosegontar commented:
`# @unittest.skip('skipping')` is removed from the test functions that I added. I wasn't using it, but included it to match the other test methods in that class. I left the original one that was on `test_search_by_tag`, but can remove if desired.

I've added a test to `test_bukuDb.py` to test the db query: `test_search_by_tag_query`. 

The tag preparation in `search_by_tag` is now in a `prep_tag_search` helper function, and there is a corresponding  `test_prep_tag_search` test in `test_buku.py`.

rachmadaniHaryono commented on tests/test_buku.py:
```diff
@@ -107,6 +107,31 @@ def test_parse_tags(keywords, exp_res):
     assert res == exp_res
 
 
+def test_prep_tag_search():
```
Use pytest paramerize here

rachmadaniHaryono commented:
I think the change on that test using pytest 's paramerize  is the last one.

@jarun  If I didn't respond, you can merge the pr

mosegontar commented:
Updated!

jarun commented:
Thanks guys!
